### ISSUE TITLE
cdist gradient computation is broken

### ISSUE THREAD

**Issue by Balandat** [pytorch/pytorch#26076]
## 🐛 Bug

Gradient computation of `torch.cdist` results in non-deterministic behavior in some simple cases. It appears there is some kind of overflow or similar going on. 

## To Reproduce

```
def test_grad():
    x1 = torch.tensor([[0.]], requires_grad=True)
    x2 = torch.tensor([[0.5], [1.0]], requires_grad=True)
    res = torch.cdist(x1, x2)
    res[0, 0].backward()
    print(x1.grad, x2.grad)
```

This produces the following (depending on the setting (pytorch, OS) this happens from almost every time to only occasionally).
```
>>> test_grad()
tensor([[-1.]]) tensor([[1.0000e+00], [1.4013e-45]])

>>> test_grad()
tensor([[-1.]]) tensor([[1.], [0.]])

>>> test_grad()
tensor([[-1.]]) tensor([[1.0000e+00], [1.1285e+07]])
```

## Expected behavior

This should be deterministic and correct.

## Environment

This happens on both pytorch 1.2.0 and on pytorch master, on OSX as well as on Linux. 

## Additional context

There are other issues with `cdist` that could be related (in particular the second one):
#25799
#24345

cc @ezyang @gchanan @zou3519

**Comment by Balandat** [comment]
@jacobrgardner, @gpleiss

**Comment by ngimel** [comment]
This happens when `x1` has a size `1xn`. Then the result and the incoming gradient have the size `1xm` (where m is the number of vectors in x2), `grad.t()` is reported as contiguous, but it's last stride is still `m`, and computing offsets based on the last stride here becomes incorrect https://github.com/pytorch/pytorch/blob/fbf991d06279ea6828ec133b186dbef3bc16522b/aten/src/ATen/native/cpu/DistanceOpsKernel.cpp#L392 (`grad_k += gs`).
